### PR TITLE
feat(FEC-12113): Add Side Panels Management to player UI

### PR DESCRIPTION
**Pull request by JonathanTGold**
### Description of the Changes

Add Side Panels Management to player UI

Related pr: 
https://github.com/kaltura/playkit-js-ui/pull/674
https://github.com/kaltura/kaltura-player-js/pull/544
https://github.com/kaltura/playkit-js-ui-managers/pull/2

solves: FEC-12113

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/services/side-panels/side-panel-item.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  constructor(item: SidePanelItem) {
+    this.renderIcon = item.renderIcon;
+    this.renderContent = item.renderContent;
+    this.presets = item.presets || Object.values(ReservedPresetNames);
+    this.position = item.position;
+    this.expandMode = item.expandMode;
+    this.hooks = {
+      onActivate: (): void => {
+        return;
```
Comment (SivanA-Kaltura) [inline src/services/side-panels/side-panel-item.ts:26]: () => {} would be shorter

File: src/ui-managers.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  static defaultConfig = {};
+
+  constructor(name: string, player: any) {
+    super(name, player);
+
+    const sidePanelsManager: SidePanelsManager = new SidePanelsManager(player);
+    player.registerService('sidePanelsManager', sidePanelsManager);
+  }
+
+  reset(): void {
```
Comment (SivanA-Kaltura) [inline src/ui-managers.ts:15]: why do you override it if it's empty ?

Comment (JonathanTGold) [reply src/ui-managers.ts:15]: Unfortunately i have no choice due to the linter
@typescript-eslint/no-empty-function

File: src/services/side-panels/side-panel-item.ts (hunk 1 above)
Comment (JonathanTGold) [reply src/services/side-panels/side-panel-item.ts:26]: Unfortunately i have no choice due to the linter
@typescript-eslint/no-empty-function

File: src/services/side-panels/side-panels-manager.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    });
+  }
+
+  private injectComponent(
+    PanelComponent: FunctionalComponent,
+    position: SidePanelPosition
+  ): {componentRef: RefObject<Toggle>; removeComponentFunc: () => void} {
+    const componentRef: RefObject<Toggle> = createRef();
+    const removeComponentFunc = this.player.ui.addComponent({
+      label: 'SidePanelLeftDynamicComponent',
```
Comment (OrenMe) [inline src/services/side-panels/side-panels-manager.tsx:124]: I guess this name should be dynamic and not fixed(left?)

Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx]: fixed

File: src/ui-managers.ts (hunk 2 above)
Comment (JonathanTGold) [reply src/ui-managers.ts:13]: idn

File: demo/index.html [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <title>UI Managers</title>
+    <link rel="stylesheet" type="text/css" href="style.css" />
+<!--    <script src="https://cdnapisec.kaltura.com/p/2215841/embedPlaykitJs/uiconf_id/42237501/"></script>-->
+    <script src="http://localhost:8080/kaltura-ovp-player.js"></script>
```
Comment (yairans) [inline demo/index.html:8]: remove

File: demo/index.html [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    <script src="./ui-manager.js"></script>
+  </head>
+  <body>
+    <div id="player-placeholder" style="width: 640px; height: 360px"></div>
+    <script type="module">
+      import './some-plugin.js'
```
Comment (yairans) [inline demo/index.html:14]: will work w/o webpack?

File: demo/index.html [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        plugins: {
+          somePlugin: {},
+          uiManagers: {},
+        },
+      };
+
+      const player = KalturaPlayer.setup(config);
+      player.loadMedia({entryId: '1_ebs5e9cy'});
+      // player.loadMedia({entryId: '1_mwl8vadb'});
```
Comment (yairans) [inline demo/index.html:31]: remove

File: demo/some-plugin.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    const aa = this.player.getService('sidePanelsManager').addItem({
+      label: 'A',
+      expandMode: 'alongside',
+      renderIcon: IconComponent,
+      position: 'left',
+      renderContent: PanelItemComponent,
+    });
+
+    // eslint-disable-next-line no-console
+    console.log('1212', aa);
```
Comment (yairans) [inline demo/some-plugin.js:25]: remove

**Review by yairans — CHANGES_REQUESTED** [review]

File: demo/index.html (hunk 5 above)
Comment (JonathanTGold) [reply demo/index.html:14]: yes it does regardless of Webpack since the script it type module

File: demo/index.html (hunk 4 above)
Comment (JonathanTGold) [reply demo/index.html]: removed

File: demo/index.html (hunk 6 above)
Comment (JonathanTGold) [reply demo/index.html]: removed

File: src/services/side-panels/side-panels-manager.tsx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    const {id, item} = panelItemData;
+    const IconComponent: ComponentClass | FunctionalComponent = item.renderIcon!;
+    const togglePanelFunc: () => void = () => this.toggle(id);
+    this.player.ui.addComponent({
+      label: `Side-Panel-Icon-${item.label}`,
+      presets: [ReservedPresetNames.Playback, ReservedPresetNames.Live],
+      area: ReservedPresetAreas.TopBarRightControls,
+      get: function MyComponent() {
+        return (
+          <div onClick={togglePanelFunc}>
```
Comment (yairans) [inline src/services/side-panels/side-panels-manager.tsx:120]: Does the click continue to the icon component by default? if not need to pass it

File: src/services/side-panels/side-panels-manager.tsx [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  public deactivateItem(itemId: number): void {
+    const item: ItemMetadata | undefined = this.componentsRegistry.get(itemId);
+    if (item) {
+      const {position} = item.item;
+      this.activePanels[position]?.componentRef.current?.toggle();
+      this.collapse(position);
+      this.activePanels[position] = null;
+      item.item.hooks?.onDeactivate?.();
```
Comment (yairans) [inline src/services/side-panels/side-panels-manager.tsx:83]: same

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/ui-managers.ts (hunk 2 above)
Comment (JonathanTGold) [reply src/ui-managers.ts:13]: done

File: src/services/side-panels/side-panel-item-dto.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import {ComponentClass, FunctionalComponent} from 'preact';
+import {ui} from 'kaltura-player-js';
+import {ReservedPresetName, SidePanelPosition, SidePanelMode} from './types/types';
+const {ReservedPresetNames} = ui;
+
+export class SidePanelItemDto {
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panel-item-dto.ts:6]: Data Transfer Object is an object that carries data between processes
https://stackoverflow.com/questions/1051182/what-is-a-data-transfer-object-dto

File: src/services/side-panels/types/types.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+export type SidePanelPosition = 'top' | 'bottom' | 'right' | 'left';
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/types/types.ts:1]: this is just a type declaration for type Support (the ui exports the actual values, which i can’t use for  typing
the ideal situation is ui would have been  written in typescript too and exported the type too

File: webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    client: {
+      progress: true,
+    },
+  },
+};
+
+module.exports = (env, {mode}) => {
+  if (mode === 'development') {
+    config.devtool = 'source-map';
+    // config.entry = {'ui-manager': './src/index.ts',
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply webpack.config.js]: done

File: src/services/side-panels/side-panels-manager.tsx [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+      const newPanelItem: SidePanelItemDto = new SidePanelItemDto(item);
+      const {componentRef, removeComponentFunc} = this.injectPanelComponent(item);
+      const newItemMetadata: ItemMetadata = new ItemMetadata(newPanelItem, componentRef, removeComponentFunc);
+      if (item.renderIcon) this.injectIconComponent(newItemMetadata);
+      this.componentsRegistry.set(newItemMetadata.id, newItemMetadata);
+      this.logger.debug('1234 New Panel Item Added', item);
+      return newItemMetadata.id;
+    }
+    this.logger.error('invalid SidePanelItem parameters', item);
+    throw new Error('invalid SidePanelItem parameters');
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:41]: i think usually when you send wrong parameter to method you expect en error

File: src/services/side-panels/side-panel-item-dto.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  readonly label: string;
+  readonly renderIcon?: ComponentClass | FunctionalComponent;
+  readonly renderContent: ComponentClass | FunctionalComponent;
+  readonly presets?: ReservedPresetName[];
+  readonly position: SidePanelPosition;
+  readonly expandMode: SidePanelMode;
+  readonly hooks?: {onActivate: (() => void) | null; onDeactivate: (() => void) | null};
+
+  constructor(item: SidePanelItemDto) {
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panel-item-dto.ts:15]: fixed

File: src/services/side-panels/side-panels-manager.tsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+      // Deactivate the opposite panel if is active
+      const oppositePosition: SidePanelPosition = SidePanelsManager.getOppositePanelPosition(position);
+      if (this.activePanels[oppositePosition]) {
+        this.deactivateItem(this.activePanels[oppositePosition]!.id);
+      }
+      // Update new item as active
+      item.componentRef.current?.toggle();
+      this.expand(position, expandMode);
+      this.activePanels[position] = item;
+      item.item.hooks?.onActivate?.();
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:72]: fixed

File: src/services/side-panels/side-panels-manager.tsx (hunk 9 above)
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:83]: fixed

File: src/services/side-panels/side-panel-item-dto.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+const {ReservedPresetNames} = ui;
+
+export class SidePanelItemDto {
+  readonly label: string;
+  readonly renderIcon?: ComponentClass | FunctionalComponent;
+  readonly renderContent: ComponentClass | FunctionalComponent;
+  readonly presets?: ReservedPresetName[];
+  readonly position: SidePanelPosition;
+  readonly expandMode: SidePanelMode;
+  readonly hooks?: {onActivate: (() => void) | null; onDeactivate: (() => void) | null};
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panel-item-dto.ts:13]: done

File: src/services/side-panels/item-metadata.ts [hunk 17]
```diff
@@ -0,0 +1,17 @@
+import {RefObject} from 'preact';
+import {Toggle} from './ui/side-panel.component';
+import {SidePanelItemDto} from './side-panel-item-dto';
+
+export class ItemMetadata {
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/item-metadata.ts:5]: done

File: src/services/side-panels/side-panels-manager.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (yairans) [reply src/services/side-panels/side-panels-manager.tsx:46]: don't think so. for the app it's not an error

File: src/services/side-panels/side-panels-manager.tsx [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+  public isItemActive(itemId: number): boolean {
+    const itemMetadata: ItemWrapper | undefined = this.componentsRegistry.get(itemId);
+    return itemMetadata ? this.activePanels[itemMetadata.item.position]?.id === itemId : false;
+  }
+
+  public reset(): void {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    for (const value of this.componentsRegistry.values()) {
+      this.deactivateItem(value.id);
```
Comment (yairans) [inline src/services/side-panels/side-panels-manager.tsx:101]: I think in change media the items should be totally removed

File: src/services/side-panels/side-panels-manager.tsx [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    this.logger.warn('invalid SidePanelItem parameters', item);
+    throw new Error('invalid SidePanelItem parameters');
+  }
+
+  public removeItem(itemId: number): void {
+    const item: ItemWrapper | undefined = this.componentsRegistry.get(itemId);
+    if (item) {
+      this.deactivateItem(itemId);
+      item.removeComponentFunc();
```
Comment (yairans) [inline src/services/side-panels/side-panels-manager.tsx:53]: whats about the icon? where is its removal?

File: src/services/side-panels/side-panel-item-dto.ts [hunk 20]
```diff
@@ -3,22 +3,35 @@ import {ui} from 'kaltura-player-js';
...
   readonly renderContent: ComponentClass | FunctionalComponent;
   readonly presets?: ReservedPresetName[];
   readonly position: SidePanelPosition;
   readonly expandMode: SidePanelMode;
-  readonly hooks?: {onActivate: (() => void) | null; onDeactivate: (() => void) | null};
+  readonly onActivate?: () => void;
+  readonly onDeactivate?: () => void;
+}
+
+export class SidePanelItemDto implements ISidePanelItemDto {
```
Comment (yairans) [inline src/services/side-panels/side-panel-item-dto.ts:17]: I think ISidePanelItem it's clearer

File: src/services/side-panels/side-panels-manager.tsx (hunk 18 above)
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:94]: you mean to the ui components ? call to removeComponent ?

File: src/services/side-panels/side-panels-manager.tsx (hunk 19 above)
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:48]: fixed

File: src/services/side-panels/side-panel-item-dto.ts (hunk 20 above)
Comment (JonathanTGold) [reply src/services/side-panels/side-panel-item-dto.ts]: Already fixed, the interface name too ?
i did it easily distinguish between internal type to the external  type used outside

File: src/services/side-panels/side-panels-manager.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:41]: fixed

File: src/services/side-panels/side-panels-manager.tsx [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+
+  private injectIconComponent(panelItemData: ItemWrapper): () => void {
+    const {id, item} = panelItemData;
+    const IconComponent: ComponentClass | FunctionalComponent = item.renderIcon!;
+    const togglePanelFunc: () => void = () => this.toggle(id);
+    return this.player.ui.addComponent({
+      label: `Side-Panel-Icon-${item.label}`,
+      presets: [ReservedPresetNames.Playback, ReservedPresetNames.Live],
+      area: ReservedPresetAreas.TopBarRightControls,
+      get: function MyComponent() {
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services/side-panels/side-panels-manager.tsx:126]: yes it does work once you have tabindex inside a panel even you don’t  in the panel itself

File: src/services/side-panels/side-panels-manager.tsx [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+          </Toggle>
+        );
+      },
+    });
+
+    return {componentRef, removeComponentFunc};
+  }
+
+  private static getPanelArea(position: string): string {
+    return `SidePanel${position.charAt(0).toUpperCase()}${position.slice(1)}`;
```
[the remark this answers is not in the document]
Comment (yairans) [reply src/services/side-panels/side-panels-manager.tsx:160]: whats about this one?

File: src/services/side-panels/types/types.ts [hunk 23]
```diff
@@ -0,0 +1,3 @@
+export type SidePanelPosition = 'SidePanelTop' | 'SidePanelLeft' | 'SidePanelRight' | 'SidePanelBottom';
```
Comment (yairans) [inline src/services/side-panels/types/types.ts:1]: doesn't this make getPanelArea redundent?

File: src/services/side-panels/ui/side-panel.component.tsx [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+ */
+export class Toggle extends Component<ToggleProps, ToggleState> {
+  constructor() {
+    super();
+    this.state = {on: false};
+  }
+
+  toggle(): void {
+    this.setState((state: Readonly<ToggleState>) => {
+      return {on: !state.on};
```
Comment (yairans) [inline src/services/side-panels/ui/side-panel.component.tsx:23]: whats about the animation issue on deactive?

File: src/services-manager.ts [hunk 25]
```diff
@@ -0,0 +1,29 @@
+import {BasePlugin, KalturaPlayer} from 'kaltura-player-js';
+import {SidePanelsManager} from './services/side-panels/side-panels-manager';
+
+export const pluginName = 'uiManagers';
+
+/**
+ * manages the registration of UI services
+ * @internal
+ */
+export class UIManagersPlugin extends BasePlugin {
```
[the remark this answers is not in the document]
Comment (JonathanTGold) [reply src/services-manager.ts:10]: done

File: src/services/side-panels/types/types.ts (hunk 23 above)
Comment (JonathanTGold) [reply src/services/side-panels/types/types.ts:1]: i want back to what it was before

File: webpack.config.js [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+        test: /\.tsx?$/,
+        use: 'ts-loader',
+        exclude: /node_modules/,
+      },
+      {
+        test: /\.scss/,
+        use: [
+          'style-loader',
+          {
+            loader: "css-loader",
```
Comment (yairans) [inline webpack.config.js:18]: remove

File: src/services/side-panels/ui/toggel.scss [hunk 27]
```diff
@@ -0,0 +1,8 @@
+.side-panel {
+  transition: all;
```
Comment (yairans) [inline src/services/side-panels/ui/toggel.scss:2]:
```suggestion
  transition: visibility;
```

File: src/services/side-panels/ui/toggel.tsx [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    this.setState((state: Readonly<ToggleState>) => {
+      return { on: !state.on };
+    });
+  }
+
+  render(): ComponentChild {
+    return (
+      <div
+        className={[styles.sidePanel, this.state.on ? styles.active : ''].join(' ')}
+        style={{ transitionDuration: `${defaultTransitionTime}ms` }}
```
Comment (yairans) [inline src/services/side-panels/ui/toggel.tsx:36]: the transition should be applied from on to off only no?

File: src/services/side-panels/ui/toggel.scss [hunk 29]
```diff
@@ -0,0 +1,8 @@
...
+  visibility: hidden;
+}
+
+.active {
+  visibility: visible;
```
Comment (yairans) [inline src/services/side-panels/ui/toggel.scss:7]: transition: none (not sure this is the syntax)

Comment (JonathanTGold) [reply src/services/side-panels/ui/toggel.scss:7]: [this is the syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/visibility)

File: src/services/side-panels/ui/toggel.tsx (hunk 28 above)
Comment (JonathanTGold) [reply src/services/side-panels/ui/toggel.tsx:36]: moved transition properly  the active class

File: src/services/side-panels/ui/toggel.scss (hunk 27 above)
Comment (JonathanTGold) [reply src/services/side-panels/ui/toggel.scss]: fixed

**Review by yairans — APPROVED** [review]

**Pull request merged**